benches/wasm-kernel/src/lib.rs

```diff
@@ -48,3 +40,84 @@ pub extern "C" fn bench_tiny_keccak(test_data: *const TinyKeccakTestData) {
 		keccak.finalize((*test_data).result);
 	}
 }
+
+pub struct RevComplementTestData {
+	input: ManuallyDrop<Box<[u8]>>,
+	output: ManuallyDrop<Box<[u8]>>,
+}
+
+#[no_mangle]
+pub extern "C" fn prepare_rev_complement(size: usize) -> *mut RevComplementTestData {
+	let input = vec![0; size];
+	let output = vec![0; size];
+
+	let test_data = Box::new(
+		RevComplementTestData {
+			input: ManuallyDrop::new(input.into_boxed_slice()),
+			output: ManuallyDrop::new(output.into_boxed_slice()),
+		}
+	);
+
+	// Basically leak the pointer to the test data. This shouldn't be harmful since `prepare` is called
+	// only once per bench run (not for the iteration), and afterwards whole memory instance is discarded.
+	Box::into_raw(test_data)
+}
+
+#[no_mangle]
+pub extern "C" fn rev_complement_input_ptr(test_data: *mut RevComplementTestData) -> *mut u8 {
+	unsafe {
+		(*test_data).input.as_mut_ptr()
+	}
+}
+
+#[no_mangle]
+pub extern "C" fn rev_complement_output_ptr(test_data: *mut RevComplementTestData) -> *const u8 {
+	unsafe {
+		(*test_data).output.as_ptr()
+	}
+}
+
+#[no_mangle]
+pub extern "C" fn bench_rev_complement(test_data: *mut RevComplementTestData) {
+	unsafe {
+		let result = rev_complement::run(&*(*test_data).input);
+		(*test_data).output.copy_from_slice(&result);
+	}
+}
+
+pub struct RegexReduxTestData {
+	input: ManuallyDrop<Box<[u8]>>,
+	output: Option<usize>,
+}
+
+#[no_mangle]
+pub extern "C" fn prepare_regex_redux(size: usize) -> *mut RegexReduxTestData {
+	regex_redux::prepare();
+
+	let input = vec![0; size];
+	let test_data = Box::new(
+		RegexReduxTestData {
+			input: ManuallyDrop::new(input.into_boxed_slice()),
+			output: None,
+		}
+	);
+
+	// Basically leak the pointer to the test data. This shouldn't be harmful since `prepare` is called
+	// only once per bench run (not for the iteration), and afterwards whole memory instance is discarded.
+	Box::into_raw(test_data)
+}
+
+#[no_mangle]
+pub extern "C" fn regex_redux_input_ptr(test_data: *mut RegexReduxTestData) -> *mut u8 {
+	unsafe {
+		(*test_data).input.as_mut_ptr()
+	}
+}
+
+#[no_mangle]
+pub extern "C" fn bench_regex_redux(test_data: *mut RegexReduxTestData) {
+	unsafe {
+		let result = regex_redux::run(&*(*test_data).input);
+		(*test_data).output = Some(result);
```

Comment:
This doesn't drop the old data any more, I was thinking we avoid dropping the long-lived data but the data that we only need for one iteration (i.e. this data) gets dropped.

Reply:
Hm, are you sure? My impression was that it only doesn't drop contained data on dropping of `ManuallyDrop` itself; however, an assignment is a different case. 

I'd expect that it desugars like that:

```rust
let tmp = (*test_data).output;
drop(tmp)
(*test_data).output = new_output;
```

Reply:
Yep, can confirm that:

https://play.rust-lang.org/?gist=1f13c35e1f5f03e99757e7737e50fcb6&version=stable&mode=debug

Reply:
That's really strange, I wouldn't have expected that at all. I guess that's just how the semantics of `DerefMut` works though, so it makes sense.